### PR TITLE
added option to keep font color on selection + use consistent selection color in help (fixes: #11501)

### PR DESCRIPTION
**Pull request by toger5**
from:
screen shot 2017-09-19 at 02 59 53

to:
screen shot 2017-09-19 at 04 20 20
screen shot 2017-09-19 at 04 20 34

### CODE REVIEW

**Comment by groud** [comment]
Is it really necessary to make it an option ? I'm not sure we should make configurable that kind of things, it bloats the configuration menu for something people usually don't really care about... ^^

That being said, I like the new one, but the selection is maybe a little bit too subtle.

**Comment by toger5** [comment]
@groud
Agreed on everything. The color is to subtle + it shoukd just be the default and we get rid of selected font color.

**Comment by toger5** [comment]
So removing it to be optional is desired?
(just to make sure...)

**Comment by groud** [comment]
We can discuss that on PR tuesday

**Comment by Zylann** [comment]
I cannot see images on Github for some reason, but by reading the title, this may be what I was looking for: https://github.com/godotengine/godot/issues/11502

**Comment by toger5** [comment]
@Zylann yea it is exacrly what is done in this pr. Solves #11502.
What are your opinions on making it an option (as it is imolemented right now), or should it just be the default behaviour. No option.

**Comment by Zylann** [comment]
I thought an option would be ok since the idea is to allow dark selection in a dark theme the added bonus of preserving syntax highlighting. Current default selection color is bright, so dark override is justified, except when customized to dark too.

**Comment by djrm** [comment]
@toger5 i think this should be the default, currently there are a lot of obscure options, which i think nobody will notice anyway.

**Comment by toger5** [comment]
@groud I think we decided that it should not be 'configurable'
but since you mentioned configuration menu that also might be misunderstood.
I don't want to make it an option in the settings menu. I just want to give the text edit/ rich text label class a function for setting if it should override the font color on selection.
But in the editor this is always set + it is the default. only for personal projects there is still the option to disable it.

File: scene/gui/rich_text_label.cpp [hunk 1]
```diff
@@ -1968,6 +1982,7 @@ RichTextLabel::RichTextLabel() {
 	tab_size = 4;
 	default_align = ALIGN_LEFT;
 	underline_meta = true;
+	override_selected_font_color = true;
```
Comment (toger5) [inline scene/gui/rich_text_label.cpp]: so changing this to false to make it the default would basically make it always disabled except if someone really wants to enable it...

File: scene/gui/text_edit.cpp [hunk 2]
```diff
@@ -4861,6 +4872,7 @@ TextEdit::TextEdit() {
 	readonly = false;
 	setting_row = false;
 	draw_tabs = false;
+	override_selected_font_color = true;
```
Comment (toger5) [inline scene/gui/text_edit.cpp]: and this one to false too

**Comment by toger5** [comment]
wouldn't that be even better than just removing the option?

**Comment by djrm** [comment]
it should not be removed from RTL, this is just a theme thing, maybe this was misunderstood.

**Comment by toger5** [comment]
@djrm can you rephrase what you mean?
a theme thing? it will be a setting for the rtf (we could Make it a theme property though...) like any other setting which gets set on during setup for the according rtf and in it's initialiser.

**Comment by djrm** [comment]
@toger5 i mean it is good as an option for RTL, but not good as an option for the editor interface.

**Comment by Zylann** [comment]
RTL?

**Comment by groud** [comment]
> @groud I think we decided that it should not be 'configurable'
but since you mentioned configuration menu that also might be misunderstood.
I don't want to make it an option in the settings menu. I just want to give the text edit/ rich text label class a function for setting if it should override the font color on selection.

This is ok for me :). I just didn't wanted it to be an editor setting, but it can be in the API.

> RTL?

RichTextLabel I guess

**Comment by toger5** [comment]
Okay  great👍

**Comment by toger5** [comment]
Updated it so that keeping the color is the default.
it is still an option and can get turned on when desired.

**Comment by toger5** [comment]
also it is using a little stronger color.

**Comment by Zylann** [comment]
> also it is using a little stronger color.
What is using a stronger color? (sorry my ISP blocks Github images)

**Comment by toger5** [comment]
@Zylann 
old version: https://ibb.co/hW34ZQ

new version: https://ibb.co/nQ2yuQ

(can I you see the embedded ones? when using different host for image?)

**Comment by toger5** [comment]
@djrm 
during working on it, I found two (debateble) issues.
1. the adapt theme is not adapting the text theming color instead it's loading the colors from the theme.
2. I'm not sure if it is of any advantage to separate the helpmcolors from text editor theme colors. I think they both should use the theme colors. and when set to adaptive they are updated accordingly (light/ dark theme)

**Comment by Zylann** [comment]
@toger5 I can't see or download anything hosted on Github and no avatars. Blame Sky UK :p
Your second screenshot ("new version") looks like the screenshot in your previous post... are you sure you didn't inverted them?
I prefer the one where selection is darker indeed, but i don't understand why there are two shades of selection in the same screenshot? Did you set the same color for current line highlight and selection?

**Comment by djrm** [comment]
@toger5 yeah, after carefull thinking i decided that adaptive does not mean, that the editor color would adapt to lets say accent color, this is because besides finding colors that match might be a little bit harder, it could also be confusing for the user, since changing the theme could mess up the colors of the text_editor theme that they are used to. this is not the case with docs theme, that can it fine if it changes colors accordingly bacause with it i only care that the more relevant parts are more visible, also as i mentioned earlier, having a separate theme for docs would make highlighted code (currently unsuported) in docs look more distinguishable.

**Comment by toger5** [comment]
@djrm the reason i like having the same highlighting for help + code editor is, that it helps you to see what is a class type / function ... Since you are used to the color. 
Also awitching back and forth (help, code) with a coustom code theme is more comfortable when having the same color in both of them.
I see your concern about loosing coustom highlight settings when switching text color theme preset. But i think this should be fixed with an alert telling you that not saving the theme will result in lossing it when switching preset.

But i agree that adaptive shouldnt mean that it adapt to highlight color. (maybe it can for one or two highlightings (maybe strings could use the highlight color... Or functions ... )) it just should adapt to light/dark.

But i also dont like atm that adaptive is a compleatlt seperate mode. Whis is in the same list than other presets. Since it is not changing the values in highlighting settings section.

I think the better solution is:
Having docs + code use the code highlighting colors from the settings.
Adaptive means that based on the ligh/dark those values get updated.

**Comment by djrm** [comment]
@toger5 using accent color for some things wont work, because other static color could get confused with the color you choose, that why i think its simply not a good idea.

also yeah thats what currently adaptive means.

about the editor help, using theme color, i guess it could use the code colors for keywords, classes and such, but currently the help looks so nice (in my opinion, that this will be hard, but still im open to it).

**Comment by toger5** [comment]
Yea it looks pretty damn good.
But isnt it using the same colors than the code editor (at least almost the same colors?)
So if we map them correctly ot shouldnt look differnt?

**Comment by toger5** [comment]
I also agree that is no benifit in using the accent color in the code theme...
(was more or less jsut an idea.. But compleatly unecassary)

I know we are already drifting away from the actual pr...
BUT:
What about simply having two color presets for light and dark?
So ppl can have differnt code color themes when using a light/ a dark theme. And default is:
Light: "default light" ( or "godot light")
Dark: " default dark" (or "godot dark")

Just to make it easier to understand what is hppening.  (adaptive right now is kind of a mystery for moat users i think?)

**Comment by djrm** [comment]
@toger5 your help editor mightl look like the code editor because your accent color is blue, but if you change the accent color everything changes in the help.

Good thing about adaptive now its that it is the default, so user should notice that it adapts to white theme (but im not really sure).

**Comment by toger5** [comment]
@djrm oh i forgot its using the accent color...
What a dilemma. That is also kind of cool.

Im still in favor of havin a cleaned up code for adaptive theme. In terms of: no if statement for adaptive inside the code_editor class. I think adaptive should just use theme values to populate the highlight settings.
For now we could just repopulate the highlight settings when changing it to the light theme.

**Comment by toger5** [comment]
@djrm do you think there is any benefit in having consistency in help + code colors. Do you think switching back and forth is more comfortable with the same coloring?

**Comment by toger5** [comment]
update:
 - fixed the overlapping of the current line color with the selection color.

**Comment by djrm** [comment]
@toger5 the condition in code editor is because we still preserve the default theme, we could simply remove it and remove the condition, also no i dont think there is a benefit in having consistency in docs and code, simply because docs are not code, there is code in the docs, an im open to use code colors in those element but not the rest, it is clear that doc headers should use the accent color, because thats the color the user has choosen to be the color for reelevant content, as i said we can use code color for Types and keywords, but having it mixed up with everything was not good, fortunately now the code is more easy to modify, so it should no be hard to change that.

**Comment by toger5** [comment]
@djrm okay. sounds pretty well though through.

I also think we could remove the old default editor theme. (or maybe just provide it as godot2 in the .godot folder on installation) and make the new one the default settings so we don't have the conditional.

**Comment by Zylann** [comment]
@toger5 isn't the Godot 2 theme used as default for games though?

**Comment by toger5** [comment]
@Zylann my bad... the default code editor color theme. 
It is still the default and adapt is just overriding it's effect in the editor_help class

File: scene/gui/text_edit.cpp [hunk 3]
```diff
@@ -953,10 +950,22 @@ void TextEdit::_notification(int p_what) {
...
+						if (j == 0)
+							//first char
+							VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(0, ofs_y, (char_ofs + char_margin), get_row_height()), cache.current_line_color);
+						else if (j == str.length() - 1)
+							//last char
+							VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(char_ofs + char_margin + char_w, ofs_y, xmargin_end - (char_ofs + char_margin + char_w), get_row_height()), cache.current_line_color);
+
+						if (!in_selection)
+							VisualServer::get_singleton()->canvas_item_add_rect(ci, Rect2(Point2i(char_ofs + char_margin, ofs_y), Size2i(char_w, get_row_height())), cache.current_line_color);
+					}
```
Comment (Paulb23) [inline scene/gui/text_edit.cpp]: This will undo the behavior introduced by #11282 hiding the breakpoint marker as it will be overdrawn by the selection/current line.

Comment (toger5) [reply scene/gui/text_edit.cpp]: You are refferring to the draw order, right?
Because this changes nothing except the order and that it does not color the whole line but only the part which is not selected.

Comment (Zylann) [reply scene/gui/text_edit.cpp]: Uh, are you taking into account the fact that it could split the highlight in two if for example just one word is selected?
Also, note that the highlight should be drawn before text (as a background), so if selection color is non-transparent (also as a background), then this wouldn't be needed.

Comment (toger5) [reply scene/gui/text_edit.cpp]: @Zylann yea it is drawing two rectangles, one on the left of the selection and one on the right, if that is needed to make the whole line highlighted.
It's also drawn before (draw oder wise) or underneath (I think you get what I mean) the actual font.

Comment (Paulb23) [reply scene/gui/text_edit.cpp]: @toger5 My bad, I had a non-transparent colour so the breakpoint was hidden behind the selection. Tested master and it behaves the same, shows how often I use breakpoints :P

Comment (toger5) [reply scene/gui/text_edit.cpp]: @Paulb23 yea I also realized that they were invisible from time to time.
I can look into that in and put breakpoints on top? (drawn after the highlighting)

Oh are we talking about the breakpoints, or the debugging cursor?

Comment (Zylann) [reply scene/gui/text_edit.cpp]: I think breakpoints must be on top, as if they were characters.
Breakpoint highlight should be on top of line highlight. Well, choose a sensible order :)
